net: Use alternative port for incoming Tor connections

hebasto opened the pull request:
This PR adds ability to label incoming Tor connections as different from normal localhost connections.

Closes #8973.
Closes #16693.

Default onion service target ports are:
- 8334 on mainnnet
- 18334 on testnet
- 38334 on signet
- 18445 on regtest

To set the onion service target socket manually the extended `-bind` config option could be used:

```
$ src/bitcoind -help | grep -A 6 -e '-bind'
  -bind=<addr>[:<port>][=onion]
       Bind to given address and always listen on it (default: 0.0.0.0). Use
       [host]:port notation for IPv6. Append =onion to tag any incoming
       connections to that address and port as incoming Tor connections
       (default: 127.0.0.1:8334=onion, testnet: 127.0.0.1:18334=onion,
       signet: 127.0.0.1:38334=onion, regtest: 127.0.0.1:18445=onion)

```

Since [pr19991.02 update](https://github.com/bitcoin/bitcoin/pull/19991#issuecomment-698882284) this PR is an alternative to #19043.

hebasto committed:
net: Use network byte order for in_addr.s_addr

It is documented in the ip(7) manual page.

Saibato commented on src/net.cpp:
```diff
@@ -2300,11 +2300,14 @@ bool CConnman::InitBinds(const std::vector<CService>& binds, const std::vector<N
     if (binds.empty() && whiteBinds.empty()) {
         struct in_addr inaddr_any;
-        inaddr_any.s_addr = INADDR_ANY;
+        inaddr_any.s_addr = htonl(INADDR_ANY);
         struct in6_addr inaddr6_any = IN6ADDR_ANY_INIT;
```
nit. Why casting 0 to 0 , 0 is 0 in every byte-order? I like this to be a constant and not a hookable function.

kristapsk replied:
Is it guaranteed that `INADDR_ANY` will always be 0, on any operating system?

Saibato replied:
[rfc1122](https://tools.ietf.org/html/rfc1122#section-3.2.1.3) section-3.2.1.3

hebasto replied:
> Why casting 0 to 0 , 0 is 0 in every byte-order?

Firstly, https://www.man7.org/linux/man-pages/man7/ip.7.html:
> `in_addr` should be assigned one of the `INADDR_*` values using `htonl`...

Secondly, it is inconsistent to have `s_addr = INADDR_ANY` in a line, and `s_addr = htonl(INADDR_LOOPBACK)` in another one.

Saibato replied:
Yup i guess its just a generation thing, its a nit, but i always liked to some degree how the bitcoin client did network things different than expected, 
U will never know who wrote the example codes and how drunk or biased by a gun on there head they where at those times when the foundations  u all take for granted where coded. 
INADDR_LOOPBACK is a different case and hey what can u do if the constant is not in net order, i remember the rant of Hobbit when he bashed on Berkeley sockets implementation ( also on many other reasons) . Htonl() was one of the first predictable ways the :Gesseler had to go". And if u bind to "this host" or maybe not makes a difference.

I will not way in this more, made my point and u are in charge its now ur world, but please never trade readability for function.

sipa replied:
@saibato Are you ok?

vasild replied:
`htonl()` is the proper one even though the number is a palindrome. We had some discussion about that [here](https://github.com/bitcoin/bitcoin/pull/19628#discussion_r464156291) and [here](https://github.com/bitcoin/bitcoin/pull/19628#discussion_r473486793).

Saibato replied:
@sipa 
> @Saibato Are you ok?
<details><summary>click here :package: :tada:  . to find out!</summary>
<br>
I case u forgot,  just assign INADDR_ANY without htonl is up till now in source code right after some first commits. from November 2009 onward.
Even you later never changed that behavior and so far all run good with this.

I get nervous when things in bitcoin are changed by a drive by commit just because ...

Why even touch it?  Suppose Satoshi knew what we did in cold war and those where those magic lines that made bitcon a stable server.
I doubt that u lads are like we, but u all have the potential to grow and learn.

i like it when i can trigger u to mature, but ok is that all?  


```
    // The sockaddr_in structure specifies the address family,
    // IP address, and port for the socket that is being bound
    struct sockaddr_in sockaddr;
    memset(&sockaddr, 0, sizeof(sockaddr));
    sockaddr.sin_family = AF_INET;
    sockaddr.sin_addr.s_addr = INADDR_ANY; // bind to all IPs on this computer
    sockaddr.sin_port = DEFAULT_PORT;

```
```
commit 93cfb02acb13b74198b8e0ab72f039313ecf9f1f
Author:	s_nakamoto <s_nakamoto@1a98c847-1fd6-4fd8-948a-caf3550aa51b>  Fri Nov  6 05:50:05 2009
Committer:	s_nakamoto <s_nakamoto@1a98c847-1fd6-4fd8-948a-caf3550aa51b>  Fri Nov  6 05:50:05 2009
```
</details>

practicalswift commented:
Concept ACK

Being able to differentiate between incoming Tor connections from normal localhost connections would be very useful

glozow reviewed (commented):
Concept ACK

laanwj commented:
Concept ACK! Thank you for working on this.

vasild commented on src/init.cpp:
```diff
@@ -451,6 +451,7 @@ void SetupServerArgs(NodeContext& node)
     argsman.AddArg("-maxtimeadjustment", strprintf("Maximum allowed median peer time offset adjustment. Local perspective of time may be influenced by peers forward or backward by this amount. (default: %u seconds)", DEFAULT_MAX_TIME_ADJUSTMENT), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-maxuploadtarget=<n>", strprintf("Tries to keep outbound traffic under the given target (in MiB per 24h). Limit does not apply to peers with 'download' permission. 0 = no limit (default: %d)", DEFAULT_MAX_UPLOAD_TARGET), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-onion=<ip:port>", "Use separate SOCKS5 proxy to reach peers via Tor onion services, set -noonion to disable (default: -proxy)", ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
+    argsman.AddArg("-onionservicetargetport=<port>", strprintf("Listen for connections from Tor onion service on <port> (default: %u, testnet: %u, regtest: %u)", defaultBaseParams->OnionServiceTargetPort(), testnetBaseParams->OnionServiceTargetPort(), regtestBaseParams->OnionServiceTargetPort()), ArgsManager::ALLOW_INT | ArgsManager::NETWORK_ONLY, OptionsCategory::CONNECTION);
```
Signet not mentioned?

hebasto replied:
I think this change -- to mention signet in help output -- should be the part of another PR, no?

hebasto replied:
See #20014.

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19991#issuecomment-699936868).

vasild reviewed (commented):
* This will always open `127.0.0.1:8334`, even if the new option `-onionservicetargetport=` is not specified and the system does not use Tor in any way. I think this deserves a release notes mention.

* Binding to `127.0.0.1` would make it impossible (or hard) to run the Tor daemon on another host. What about changing to `-onionservicetarget=bind_address:port` without a default value - i.e. don't do the new bind unless `-onionservicetarget=` is specified? This would also allow to bind to an external address or to `0.0.0.0` (any).

* Notice that if another peer manages to reach and connect to the new bind without tor, then we would mistakenly think it is an incoming tor connection.

* I am just curious - what will happen if `-onionservicetargetport=8333` is specified?

hebasto commented:
@vasild
>  - This will always open `127.0.0.1:8334`, even if the new option `-onionservicetargetport=` is not specified and the system does not use Tor in any way. I think this deserves a release notes mention.

Target port opening could be dependent on `-listenonion` option value. What do you think?

>  - Binding to `127.0.0.1` would make it impossible (or hard) to run the Tor daemon on another host. What about changing to `-onionservicetarget=bind_address:port` without a default value - i.e. don't do the new bind unless `-onionservicetarget=` is specified? This would also allow to bind to an external address or to `0.0.0.0` (any).

This issue is addressed in #19043, IIUC.

>  - Notice that if another peer manages to reach and connect to the new bind without tor, then we would mistakenly think it is an incoming tor connection.

Hmmm... Need to think about this.

>  - I am just curious - what will happen if `-onionservicetargetport=8333` is specified?

It will still work but without strong distinction between Tor and clearnet, no?

naumenkogs commented:
Concept ACK

hebasto commented:
From [IRC discussion](http://www.erisian.com.au/bitcoin-core-dev/log-2020-09-23.html#l-208):
> \<vasild> what about extending the -bind=addr:port option to something like -bind=addr:port[=tor], for example: -bind=1.2.3.4:8333 -bind=1.2.3.4:8334=tor
> \<vasild> instead of adding a new option -onionservicetarget= ?
> \<wumpus> making it an attribute makes sense
> \<wumpus> (of the bind)
> \<vasild> I don't know yet how i2p works, but if it works in a similar way (the i2p daemon connects to us, like in tor) then for i2p that would mean just recognizing =i2p in addition to recognizing =tor, instead of adding new option -i2pservicetarget=addr:port
> \<wumpus> right, it's elegant way to think about it and allows for extensibility

jonatack commented:
Concept ACK. Could use test coverage. I'm running #20002 that is built on these changes and those in #19998.

hebasto committed:
net: Add alternative port for onion service

This change allows to distinguish incoming Tor connections from others.

hebasto committed:
net, refactor: Refactor CBaseChainParams::RPCPort function

hebasto committed:
refactor: Rename TorController::target to m_tor_control_center

`target` is a proper name for the onion service target address and port.
This change is required for the following commit.

hebasto committed:
net: Pass onion service target to Tor controller

hebasto commented:
@vasild @jonatack 

> what about extending the -bind=addr:port option to something like -bind=addr:port[=tor]

I like that ^ idea very much. Implemented in https://github.com/hebasto/bitcoin/commits/200924-WIP

As many other PRs are built on top of this one, I'm begging you to make some preliminary review of the suggested branch to avoid unneeded rebasing.

New version features:
- extended `-bind` configure option, now it completely replaces #19043 as well
- no extra binding when `-nolistenonion`

Also a small Q: is it better to drop doc commit cfa17e307adaec81f6036cf4f124d78f9e128c30 in favor of #19961 ?

hebasto commented:
@vasild 
>  - Notice that if another peer manages to reach and connect to the new bind without tor, then we would mistakenly think it is an incoming tor connection.

This behavior could be mitigated in #19998 by additional checking of `peer.addr` and `peer.GetAddrLocal()`.

jonatack commented:
I like that approach, too. Will look at your branch.

hebasto commented:
Updated 6e05a443bbe7090a331281290f18bedea13bab0a -> cfa17e307adaec81f6036cf4f124d78f9e128c30 ([pr19991.01](https://github.com/hebasto/bitcoin/commits/pr19991.01) -> [pr19991.02](https://github.com/hebasto/bitcoin/commits/pr19991.02), [diff](https://github.com/hebasto/bitcoin/compare/pr19991.01..pr19991.02)):

- extended `-bind` configure option (credits to @vasild), now this PR completely replaces #19043
- no extra binding when `-nolistenonion`
- added corresponding changes into `tor.md`

vasild commented on src/torcontrol.cpp:
```diff
@@ -752,7 +757,9 @@ void StartTorControl()
 
-    torControlThread = std::thread(std::bind(&TraceThread<void (*)()>, "torcontrol", &TorControlThread));
+    torControlThread = std::thread(&TraceThread<std::function<void()>>, "torcontrol", [onion_service_target] {
+        TorControlThread(onion_service_target);
```
`TorControlThread()` only takes a reference to this object and uses it in another thread. I think this is dangerous because the liftetime of the object may be shorter than the lifetime of the thread. Better make `TorControlThread()`'s argument just `CService` so that the object is copied.

hebasto replied:
wrt properties of `std::thread()` constructor and `TraceThread()` function I don't think it is dangerous. But objects of `CService` are small, so going implement passing by-copy.

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19991#issuecomment-699005931).

vasild reviewed (commented):
Just to summarize the default behavior:

### Before this PR
* If no `-bind` is given then we will bind on `0.0.0.0:8333` and `[::]:8333`.
* If `-bind=1.2.3.4:8333` is given then we will bind only on `1.2.3.4:8333`.

### With this PR
* If no `-bind` is given then we will bind on `0.0.0.0:8333`, `[::]:8333` and `127.0.0.1:8334`.
* If `-bind=1.2.3.4:8333` is given then we will bind on `1.2.3.4:8333` and on `127.0.0.1:8334` (is this ok?).
* If `-bind=1.2.3.4:8333 -bind=1.2.3.4:8334=onion` is given then we will bind on `1.2.3.4:8333` and `1.2.3.4:8334`.

hebasto commented:
@vasild 
> Just to summarize the default behavior:
> ### Before this PR
> 
>  - If no `-bind` is given then we will bind on `0.0.0.0:8333` and `[::]:8333`.
> 
>  - If `-bind=1.2.3.4:8333` is given then we will bind only on `1.2.3.4:8333`.
> 
> 
> ### With this PR
> 
>  - If no `-bind` is given then we will bind on `0.0.0.0:8333`, `[::]:8333` and `127.0.0.1:8334`.
> 
>  - If `-bind=1.2.3.4:8333` is given then we will bind on `1.2.3.4:8333` and on `127.0.0.1:8334` (is this ok?).
> 
>  -  If `-bind=1.2.3.4:8333 -bind=1.2.3.4:8334=onion` is given then we will bind on `1.2.3.4:8333` and `1.2.3.4:8334`.

The node will not bind on `127.0.0.1:8334` if non-default `listenonion=0` is provided.

vasild commented:
[This](https://lists.torproject.org/pipermail/tor-onions/2020-September/000414.html) could be an alternative way to check if an incoming connection is from tor - to ask the tor daemon if it made it. But it seems this is not possible now and even if made possible in the future then it will only be available if bitcoind has access to tor-control.

hebasto commented on src/torcontrol.cpp:
```diff
@@ -81,12 +84,12 @@ class TorControlConnection
+     * control_socket is address of the form host:port.
      * connected is the handler that is called when connection is successfully established.
      * disconnected is a handler that is called when the connection is broken.
      * Return true on success.
      */
-    bool Connect(const std::string &target, const ConnectionCB& connected, const ConnectionCB& disconnected);
+    bool Connect(const std::string& control_socket, const ConnectionCB& connected, const ConnectionCB& disconnected);
```
Maybe `control_center` (if we are already within Tor-related class)?

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19991#issuecomment-699005931).

hebasto commented on src/net.h:
```diff
@@ -201,6 +202,8 @@ class CConnman
         std::vector<NetWhitelistPermissions> vWhitelistedRange;
         std::vector<NetWhitebindPermissions> vWhiteBinds;
         std::vector<CService> vBinds;
+        bool listen_onion{true};
+        Optional<CService> onion_bind;
```
#19759

vasild replied:
Ok, drop this. I suggested it because members below use `m_` prefix. But members above don't!

hebasto replied:
https://github.com/bitcoin/bitcoin/pull/19759#issuecomment-700702839

:D

hebasto commented on src/init.cpp:
```diff
@@ -1920,13 +1917,34 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
+            }
+        } else {
+            const std::string network_type = bind_arg.substr(index + 1);
+            if (connOptions.listen_onion && network_type == "onion") {
```
```
$ src/bitcoind -bind=127.0.0.1:4242=omion | grep omion
2020-09-25T13:57:56Z [init] Command-line arg: bind="127.0.0.1:4242=omion"
Error: Cannot resolve -bind address: '127.0.0.1:4242=omion'
2020-09-25T13:57:58Z [init] Error: Cannot resolve -bind address: '127.0.0.1:4242=omion'
```

vasild replied:
Sorry, I misread the code.

hebasto commented on src/init.cpp:
```diff
@@ -1920,13 +1917,34 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
+            if (connOptions.listen_onion && network_type == "onion") {
+                const std::string truncated_bind_arg = bind_arg.substr(0, index);
+                if (Lookup(truncated_bind_arg, bind_addr, BaseParams().OnionServiceTargetPort(), false)) {
+                    connOptions.onion_bind = bind_addr;
```
> This would accept multiple =onion arguments and discard all but the last one.

That is done intentionally, leaving improving for follow ups.

vasild replied:
I think it is not good to introduce one behavior and change it later. Like "ignore all but last `-bind=addr:port=onion`" and later change it to something else because users may start to depend on the first one.

If supporting more than one `-bind=addr:port=onion` is undesired then what about emitting an error in that case?

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19991#issuecomment-700831638).

hebasto commented on src/net.cpp:
```diff
@@ -2300,11 +2303,16 @@ bool CConnman::InitBinds(const std::vector<CService>& binds, const std::vector<N
         fBound |= Bind(CService(inaddr_any, GetListenPort()), !fBound ? BF_REPORT_ERROR : BF_NONE, NetPermissionFlags::PF_NONE);
     }
+
+    if (m_listen_onion) {
+        fBound |= Bind(onion_bind, BF_EXPLICIT, NetPermissionFlags::PF_NONE);
+    }
```
Ok. I think a user should be able to explicitly disable such an additional binding? Any ideas?

hebasto replied:
Why not assign to the "-listenonion" option a new meaning "Listen incoming onion connection, and automatically create the onion service if the Tor control available". It looks pretty back-compatible, no?

vasild replied:
That would couple the two things together under one option, making it impossible to configure them independently:

N | auto create hidden service (`-listenonion`) | extra bind (`-bind=addr:port=onion`)
---|---|---
1 | no | no
2 | no | yes
3 | yes | no
4 | yes | yes

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19991#issuecomment-700831638).

hebasto commented on src/net.cpp:
```diff
@@ -2300,11 +2303,16 @@ bool CConnman::InitBinds(const std::vector<CService>& binds, const std::vector<N
+    if (m_listen_onion) {
+        fBound |= Bind(onion_bind, BF_EXPLICIT, NetPermissionFlags::PF_NONE);
+    }
+
```
For "127.0.0.1" `IsRoutable()` returns `false`, therefore it won't be passed to AddLocal(). But, in general, I agree with you.

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19991#issuecomment-699005931).

vasild replied:
Yes, this is only relevant if a routable address is supplied to `-bind=addr:port=onion`.

The added `NetPermissionFlags::PF_ONION` works, but is strange because that is not a "permission". It is just a flag telling `Bind()` "this is a covert bind address, don't advertise it". Isn't `enum BindFlags` more suitable for it? `BF_DONT_ADVERTISE` or `BF_COVERT`?

In any case - the `_ONION` in the name may be too narrow, if we want to use the same mechanism for something else in the future.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19991#issuecomment-700831638).

hebasto commented:
Updated cfa17e307adaec81f6036cf4f124d78f9e128c30 -> 0782aee0a3268e9e4de53613f229d645c2ab06b9 ([pr19991.02](https://github.com/hebasto/bitcoin/commits/pr19991.02) -> [pr19991.03](https://github.com/hebasto/bitcoin/commits/pr19991.03), [diff](https://github.com/hebasto/bitcoin/compare/pr19991.02..pr19991.03)):

- addressed some @vasild's comment

hebasto commented on src/init.cpp:
```diff
@@ -436,7 +436,7 @@ void SetupServerArgs(NodeContext& node)
     argsman.AddArg("-asmap=<file>", strprintf("Specify asn mapping used for bucketing of the peers (default: %s). Relative paths will be prefixed by the net-specific datadir location.", DEFAULT_ASMAP_FILENAME), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-bantime=<n>", strprintf("Default duration (in seconds) of manually configured bans (default: %u)", DEFAULT_MISBEHAVING_BANTIME), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
-    argsman.AddArg("-bind=<addr>", "Bind to given address and always listen on it. Use [host]:port notation for IPv6", ArgsManager::ALLOW_ANY | ArgsManager::NETWORK_ONLY, OptionsCategory::CONNECTION);
+    argsman.AddArg("-bind=<addr>", strprintf("Bind to given address and always listen on it. Use [host]:port notation for IPv6. Use =onion suffix to bind onion service target socket (default: 127.0.0.1:%u=onion, testnet: 127.0.0.1:%u=onion, regtest: 127.0.0.1:%u=onion)", defaultBaseParams->OnionServiceTargetPort(), testnetBaseParams->OnionServiceTargetPort(), regtestBaseParams->OnionServiceTarget ...
```
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19991#issuecomment-699005931).

vasild replied:
Thanks! It just occurred to me that the default for `-bind` (with this PR) is not `127.0.0.1:8334=onion` like in the description above. It is rather `-bind=0.0.0.0:8333 -bind=127.0.0.1:8334=onion`.

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19991#issuecomment-700831638).

hebasto commented:
Updated 0782aee0a3268e9e4de53613f229d645c2ab06b9 -> 20349ec2f669c6e06dfbddbcec63b1c3b6cdbc9e ([pr19991.03](https://github.com/hebasto/bitcoin/commits/pr19991.03) -> [pr19991.04](https://github.com/hebasto/bitcoin/commits/pr19991.04), [diff](https://github.com/hebasto/bitcoin/compare/pr19991.03..pr19991.04)):

- addressed some @vasild's [comment](https://github.com/bitcoin/bitcoin/pull/19991#discussion_r494957373):
> This will cause undefined behavior if `connOptions.onion_bind` is not set, are we sure it is set here?

hebasto commented on src/net.cpp:
```diff
@@ -2323,7 +2331,7 @@ bool CConnman::Start(CScheduler& scheduler, const Options& connOptions)
     }
 
-    if (fListen && !InitBinds(connOptions.vBinds, connOptions.vWhiteBinds)) {
+    if (fListen && !InitBinds(connOptions.vBinds, connOptions.vWhiteBinds, connOptions.onion_bind.get())) {
```
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19991#issuecomment-699140059).

jonatack reviewed (commented):
Approach ACK. Would it be possible to add test coverage? Perhaps similar to `feature_proxy.py`.

(Two minor comments: Perhaps separate commits like `net, refactor: Refactor CBaseChainParams::RPCPort()` to their own "no behavior change" pure refactoring PR. They lengthen review with an increased diff that isn't essential to the goal of this change. Also, commit `refactor: Rename TorController::target to m_control_socket` makes more sense when in the same commit as `Pass onion service target to Tor controller`. When reviewing the commits in order, I was wondering, "why is this being renamed?" Feel free to ignore if you disagree.)

jonatack commented on src/init.cpp:
```diff
@@ -436,7 +436,7 @@ void SetupServerArgs(NodeContext& node)
     argsman.AddArg("-asmap=<file>", strprintf("Specify asn mapping used for bucketing of the peers (default: %s). Relative paths will be prefixed by the net-specific datadir location.", DEFAULT_ASMAP_FILENAME), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-bantime=<n>", strprintf("Default duration (in seconds) of manually configured bans (default: %u)", DEFAULT_MISBEHAVING_BANTIME), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
-    argsman.AddArg("-bind=<addr>", "Bind to given address and always listen on it. Use [host]:port notation for IPv6", ArgsManager::ALLOW_ANY | ArgsManager::NETWORK_ONLY, OptionsCategory::CONNECTION);
+    argsman.AddArg("-bind=<addr>[:<port>][=onion]", strprintf("Bind to given address and always listen on it. Use [host]:port notation for IPv6. Append =onion to tag any incoming connections to that address and port as incoming tor connections (default: mainnet: 127.0.0.1:%u=onion, testnet: 127.0.0.1:%u=onion, regtest: 127.0.0.1:%u=onion)", defaultBaseParams->OnionServiceTargetPort(), testnetBase ...
```
Signet?

hebasto replied:
#20014

jonatack replied:
Yes, #20014 doesn't update `-bind`, as there's nothing to update before this PR, so it should be updated here if I'm not confused.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19991#issuecomment-699936868).

vasild replied:
> Should we warn that only the last =onion argument will be kept if more than one is passed?

Right, or even error :bomb: or support more than one `=onion`.

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19991#issuecomment-700831638).

hebasto commented on src/chainparamsbase.cpp:
```diff
@@ -40,13 +40,13 @@ const CBaseChainParams& BaseParams()
+        return MakeUnique<CBaseChainParams>("signet", 38332, 38334);
     } else if (chain == CBaseChainParams::REGTEST) {
-        return MakeUnique<CBaseChainParams>("regtest", 18443);
+        return MakeUnique<CBaseChainParams>("regtest", 18443, 18445);
```
It is an arbitrary choice. Just verified that ports are not reserved by some other service.

jonatack replied:
Thanks. I think it would be helpful to document that and say why there is an extra port for future readers of the code.

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19991#issuecomment-699936868).

hebasto commented on src/torcontrol.cpp:
```diff
@@ -536,7 +541,7 @@ void TorController::auth_cb(TorControlConnection& _conn, const TorControlReply&
         // Request onion service, redirect port.
         // Note that the 'virtual' port is always the default port to avoid decloaking nodes using other ports.
-        _conn.Command(strprintf("ADD_ONION %s Port=%i,127.0.0.1:%i", private_key, Params().GetDefaultPort(), GetListenPort()),
+        _conn.Command(strprintf("ADD_ONION %s Port=%i,%s", private_key, Params().GetDefaultPort(), m_target.ToStringIPPort()),
```
We already have the following log message:
```
2020-09-28T10:51:00Z [init] Bound to 127.0.0.1:18334
```

jonatack replied:
Oh, quite right, I was grepping on `-ni bind` ;/

master
```
456006:2020-09-28T10:33:52Z Bound to [::]:8333
456007:2020-09-28T10:33:52Z Bound to 0.0.0.0:8333
```
this branch
```
456232:2020-09-28T11:16:58Z Bound to [::]:8333
456233:2020-09-28T11:16:58Z Bound to 0.0.0.0:8333
456234:2020-09-28T11:16:58Z Bound to 127.0.0.1:8334
```

hebasto commented:
Updated 20349ec2f669c6e06dfbddbcec63b1c3b6cdbc9e -> f1e8a24436b0110d36a56a17909ab7bd9b67899f ([pr19991.04](https://github.com/hebasto/bitcoin/commits/pr19991.04) -> [pr19991.05](https://github.com/hebasto/bitcoin/commits/pr19991.05), [diff](https://github.com/hebasto/bitcoin/compare/pr19991.04..pr19991.05)):

- addressed @jonatack's comments

hebasto commented on src/torcontrol.cpp:
```diff
@@ -193,16 +196,16 @@ void TorControlConnection::eventcb(struct bufferevent *bev, short what, void *ct
 }
 
-bool TorControlConnection::Connect(const std::string &target, const ConnectionCB& _connected, const ConnectionCB&  _disconnected)
+bool TorControlConnection::Connect(const std::string& tor_control_center, const ConnectionCB& _connected, const ConnectionCB&  _disconnected)
```
[Updated](https://github.com/bitcoin/bitcoin/pull/19991#issuecomment-699936868).

hebasto commented:
Updated 20349ec2f669c6e06dfbddbcec63b1c3b6cdbc9e -> d0e349140fac872ebf01c3826528a84bd6a89519 ([pr19991.04](https://github.com/hebasto/bitcoin/commits/pr19991.04) -> [pr19991.06](https://github.com/hebasto/bitcoin/commits/pr19991.06), [diff](https://github.com/hebasto/bitcoin/compare/pr19991.04..pr19991.06)):

- addressed @jonatack's comments

@jonatack 
> When reviewing the commits in order, I was wondering, "why is this being renamed?"

Commit message updated.

hebasto commented:
@jonatack 
> Would it be possible to add test coverage? Perhaps similar to `feature_proxy.py`.

`feature_proxy.py` tests outbound connections, IIUC.

hebasto commented:
Rebased d0e349140fac872ebf01c3826528a84bd6a89519 -> efa45f5fde37ebf20f318da74d53f056fd71d69f ([pr19991.06](https://github.com/hebasto/bitcoin/commits/pr19991.06) -> [pr19991.07](https://github.com/hebasto/bitcoin/commits/pr19991.07)) due to the conflict with #20014.

hebasto committed:
net, refactor: Move AddLocal call one level up

This change simplifies the following commit.

hebasto committed:
net: Extend -bind config option with optional network type

hebasto committed:
doc: Update onion service target port numbers in tor.md

laanwj commented:
Tested ACK. I've verified that this opens a new port (8334, bound on 127.0.0.1 only) and that `torcontrol` makes incoming connection go to there. This allows to distinguish incoming Tor connections. This works together with #19954.

hebasto commented:
Updated efa45f5fde37ebf20f318da74d53f056fd71d69f -> 30f2355d96bf67a30a06c1fcf6d27d242b05491b ([pr19991.07](https://github.com/hebasto/bitcoin/commits/pr19991.07) -> [pr19991.09](https://github.com/hebasto/bitcoin/commits/pr19991.09), [diff](https://github.com/hebasto/bitcoin/compare/pr19991.07..pr19991.09)).

1. Addressed @vasild's [comment](https://github.com/bitcoin/bitcoin/pull/19991#discussion_r494948385):
> Should we support more than one onion bind address? Maybe it would make sense on a machine with many interfaces.

and @jonatack's [comment](https://github.com/bitcoin/bitcoin/pull/19991#discussion_r495769826):
> Should we warn that only the last `=onion` argument will be kept if more than one is passed?

If `-listenonion=1` is provided, the node emits a warning about multiple bind addresses.

2. Addressed @vasild's [comment](https://github.com/bitcoin/bitcoin/pull/19991#discussion_r494961073):
> I don't think the new functionality should be tied to the `-listenonion` option.

3. Addressed @vasild's [comment](https://github.com/bitcoin/bitcoin/pull/19991#discussion_r496743497):
> Isn't `enum BindFlags` more suitable for it? `BF_DONT_ADVERTISE` or `BF_COVERT`?

4. Addressed @vasild's [comment](https://github.com/bitcoin/bitcoin/pull/19991#discussion_r496689477):
> Thanks! It just occurred to me that the default for `-bind` (with this PR) is not `127.0.0.1:8334=onion` like in the description above. It is rather `-bind=0.0.0.0:8333 -bind=127.0.0.1:8334=onion`.

laanwj commented on src/net.cpp:
```diff
@@ -20,6 +20,7 @@
 #include <protocol.h>
 #include <random.h>
 #include <scheduler.h>
+#include <torcontrol.h>
```
I really hope we can avoid introducing this circular dependency.

hebasto commented:
Updated 30f2355d96bf67a30a06c1fcf6d27d242b05491b -> 3c24f75d4ce7d3ce1e30d334ea30873a15633e46 ([pr19991.09](https://github.com/hebasto/bitcoin/commits/pr19991.09) -> [pr19991.10](https://github.com/hebasto/bitcoin/commits/pr19991.10), [diff](https://github.com/hebasto/bitcoin/compare/pr19991.09..pr19991.10)):

- addressed @laanwj's [comment](https://github.com/bitcoin/bitcoin/pull/19991#discussion_r497391823):
> I really hope we can avoid introducing this circular dependency.

laanwj commented:
Thanks for addressing my comment. 

Re-ACK 3c24f75d4ce7d3ce1e30d334ea30873a15633e46

hebasto commented:
Updated 3c24f75d4ce7d3ce1e30d334ea30873a15633e46 -> acba1851ef78d88c2442813174c2b559047b61e8 ([pr19991.10](https://github.com/hebasto/bitcoin/commits/pr19991.10) -> [pr19991.11](https://github.com/hebasto/bitcoin/commits/pr19991.11), [diff](https://github.com/hebasto/bitcoin/compare/pr19991.10..pr19991.11)):

- type of `Options.onion_binds` switched from `std::vector` to `std::set` that is more appropriate, and makes the following #19998 easier.

Sjors commented on src/net.cpp:
```diff
@@ -84,6 +84,7 @@ enum BindFlags {
     BF_NONE         = 0,
     BF_EXPLICIT     = (1U << 0),
     BF_REPORT_ERROR = (1U << 1),
+    BF_DONT_ADVERTISE = (1U << 2),
```
These flags would benefit from some documentation.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19991#issuecomment-702253099).

Sjors reviewed (approved):
tACK 3c24f75d4ce7d3ce1e30d334ea30873a15633e46 though I'm not super familiar with some of this networking code

I was able to automagically spin up a hidden service that listened on "virtual" 8333 and binds to 8334 (as well as a custom 7000). And I was able to connect another peer to it.

jonatack commented:
Will try to propose some test coverage and review asap.

vasild reviewed (approved):
ACK 3c24f75d4

A few nits and a question below, feel free to ignore.

This PR will cause an extra bind on `127.0.0.1:8334` by default. I think this deserves a release notes mention. There is no way to switch that off (one can only change the address and/or port), I think that is ok. If it is a concern, then maybe we can tweak it to: if `-bind=` is explicitly specified then don't do an extra bind and redirect the automatically created tor service (if we create one) to the normal port (8333 or whatever is given to `-bind`).

hebasto commented:
Updated 3c24f75d4ce7d3ce1e30d334ea30873a15633e46 -> 96571b3d4cb4cda0fd3d5a457ae4a12f615de82b ([pr19991.10](https://github.com/hebasto/bitcoin/commits/pr19991.10) -> [pr19991.12](https://github.com/hebasto/bitcoin/commits/pr19991.12), [diff](https://github.com/hebasto/bitcoin/compare/pr19991.10..pr19991.12)):

- addressed all @Sjors' and @vasild's comments

hebasto commented on src/init.cpp:
```diff
@@ -1940,13 +1937,39 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
+    if (args.GetBoolArg("-listenonion", DEFAULT_LISTEN_ONION)) {
+        const auto bind_addr = connOptions.onion_binds.front();
+        if (connOptions.onion_binds.size() > 1) {
+            InitWarning(strprintf(_("More than one bind address is provided for automatic Tor onion service. Using %s"), bind_addr.ToStringIPPort()));
```
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19991#issuecomment-702253099).

hebasto commented on src/net.cpp:
```diff
@@ -2308,10 +2305,18 @@ bool CConnman::Bind(const CService &addr, unsigned int flags, NetPermissionFlags
+bool CConnman::InitBinds(
+    const std::vector<CService>& binds,
+    const std::vector<NetWhitebindPermissions>& whiteBinds,
+    std::vector<CService> onion_binds)
```
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19991#issuecomment-702253099).

hebasto commented on test/functional/test_framework/test_framework.py:
```diff
@@ -436,7 +436,7 @@ def get_bin_from_version(version, bin_name, bin_default):
             extra_confs = [["bind=127.0.0.1"]] * num_nodes
         else:
-            extra_confs = [[]] * num_nodes
+            extra_confs = [["listenonion=0"]] * num_nodes
```
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19991#issuecomment-702253099).

Sjors commented:
re-utACK 96571b3d4cb4cda0fd3d5a457ae4a12f615de82b

hebasto commented:
@laanwj Assign [0.21.0](https://github.com/bitcoin/bitcoin/milestone/45) milestone?

vasild commented on src/net.cpp:
```diff
@@ -2322,11 +2331,16 @@ bool CConnman::InitBinds(const std::vector<CService>& binds, const std::vector<N
     }
+
+    for (const auto& addr_bind : onion_binds) {
+        fBound |= Bind(addr_bind, BF_EXPLICIT | BF_DONT_ADVERTISE, NetPermissionFlags::PF_NONE);
```
nit: maybe `BF_REPORT_ERROR` should be here too, like with `binds` and `whiteBinds`?

vasild reviewed (approved):
ACK 96571b3d4

Looks good, I just wonder if we can do even better. The following are just some thoughts, feel free to ignore them.

The semantic of `-bind` is to restrict the binding only to some address. If not specified, then the user does not care and we bind to `0.0.0.0`, if specified then the user cares and we bind only to the given address.

With this PR, if no `-bind` is given then we would bind to `0.0.0.0` and to `127.0.0.1` which is ok (the user does not care to restrict the binding). However:

* If only `-bind=addr` is given (without `-bind=...=onion`) then we would bind to `addr` _and_ to `127.0.0.1` in addition which may be unexpected assuming the perception of `-bind=addr` is "bind _only_ to `addr`".

* If only `-bind=addr=onion` is given (without ordinary `-bind=...`) then we would bind to `addr` _and_ to `0.0.0.0` in addition.

What about changing the above two cases to not do the additional bind? I.e. if `-bind` is given then only bind to the specified address:

* If only `-bind=addr` is given (without `-bind=...=onion`) then bind to `addr` (only). If we are creating tor hidden service then use `addr` as target. Same behavior as before this PR.

* If only `-bind=addr=onion` is given (without ordinary `-bind=...`) then bind to `addr` (only), consider incoming connections as tor and do not advertise `addr`. I.e. a tor-only node.

<details>
<summary>Example patch to achieve the above</summary>

```diff
diff --git i/src/init.cpp w/src/init.cpp
index 773e23d08..d1fed90ae 100644
--- i/src/init.cpp
+++ w/src/init.cpp
@@ -1955,31 +1955,39 @@ bool AppInitMain(const util::Ref& context, NodeContext& node, interfaces::BlockA
                 }
             }
         }
         return InitError(ResolveErrMsg("bind", bind_arg));
     }
 
-    if (connOptions.onion_binds.empty()) {
-        connOptions.onion_binds.push_back(DefaultOnionServiceTarget());
-    }
-
-    if (args.GetBoolArg("-listenonion", DEFAULT_LISTEN_ONION)) {
-        const auto bind_addr = connOptions.onion_binds.front();
-        if (connOptions.onion_binds.size() > 1) {
-            InitWarning(strprintf(_("More than one onion bind address is provided. Using %s for the automatically created Tor onion service."), bind_addr.ToStringIPPort()));
-        }
-        StartTorControl(bind_addr);
-    }
-
     for (const std::string& strBind : args.GetArgs("-whitebind")) {
         NetWhitebindPermissions whitebind;
         bilingual_str error;
         if (!NetWhitebindPermissions::TryParse(strBind, whitebind, error)) return InitError(error);
         connOptions.vWhiteBinds.push_back(whitebind);
     }
 
+    if (args.GetBoolArg("-listenonion", DEFAULT_LISTEN_ONION)) {
+        CService onion_service_target;
+        if (!connOptions.onion_binds.empty()) {
+            onion_service_target = connOptions.onion_binds.front();
+            if (connOptions.onion_binds.size() > 1) {
+                InitWarning(
+                    strprintf(_("More than one onion bind address is provided. Using %s "
+                                "for the automatically created Tor onion service."),
+                              onion_service_target.ToStringIPPort()));
+            }
+        } else if (!connOptions.vBinds.empty()) {
+            onion_service_target = connOptions.vBinds.front();
+        } else if (!connOptions.vWhiteBinds.empty()) {
+            onion_service_target = connOptions.vWhiteBinds.front().m_service;
+        } else {
+            onion_service_target = DefaultOnionServiceTarget();
+        }
+        StartTorControl(onion_service_target);
+    }
+
     for (const auto& net : args.GetArgs("-whitelist")) {
         NetWhitelistPermissions subnet;
         bilingual_str error;
         if (!NetWhitelistPermissions::TryParse(net, subnet, error)) return InitError(error);
         connOptions.vWhitelistedRange.push_back(subnet);
     }
diff --git i/src/net.cpp w/src/net.cpp
index 7f575b66c..2f67d07ed 100644
--- i/src/net.cpp
+++ w/src/net.cpp
@@ -17,12 +17,13 @@
 #include <net_permissions.h>
 #include <netbase.h>
 #include <node/ui_interface.h>
 #include <protocol.h>
 #include <random.h>
 #include <scheduler.h>
+#include <torcontrol.h>
 #include <util/strencodings.h>
 #include <util/translation.h>
 
 #ifdef WIN32
 #include <string.h>
 #else
@@ -2326,24 +2327,24 @@ bool CConnman::InitBinds(
     for (const auto& addrBind : binds) {
         fBound |= Bind(addrBind, (BF_EXPLICIT | BF_REPORT_ERROR), NetPermissionFlags::PF_NONE);
     }
     for (const auto& addrBind : whiteBinds) {
         fBound |= Bind(addrBind.m_service, (BF_EXPLICIT | BF_REPORT_ERROR), addrBind.m_flags);
     }
-    if (binds.empty() && whiteBinds.empty()) {
+    for (const auto& addr_bind : onion_binds) {
+        fBound |= Bind(addr_bind, BF_EXPLICIT | BF_REPORT_ERROR | BF_DONT_ADVERTISE, NetPermissionFlags::PF_NONE);
+    }
+    if (binds.empty() && whiteBinds.empty() && onion_binds.empty()) {
         struct in_addr inaddr_any;
         inaddr_any.s_addr = htonl(INADDR_ANY);
         struct in6_addr inaddr6_any = IN6ADDR_ANY_INIT;
         fBound |= Bind(CService(inaddr6_any, GetListenPort()), BF_NONE, NetPermissionFlags::PF_NONE);
         fBound |= Bind(CService(inaddr_any, GetListenPort()), !fBound ? BF_REPORT_ERROR : BF_NONE, NetPermissionFlags::PF_NONE);
-    }
 
-    for (const auto& addr_bind : onion_binds) {
-        fBound |= Bind(addr_bind, BF_EXPLICIT | BF_DONT_ADVERTISE, NetPermissionFlags::PF_NONE);
+        fBound |= Bind(DefaultOnionServiceTarget(), BF_NONE | BF_DONT_ADVERTISE, NetPermissionFlags::PF_NONE);
     }
-
     return fBound;
 }
 
 bool CConnman::Start(CScheduler& scheduler, const Options& connOptions)
 {
     Init(connOptions);
```
</details>

hebasto commented:
@vasild Thanks for your review and suggestion! It seems your patch introduces a new circular dependency. Could we leave improvements for follow ups wrt coming feature freeze deadline?

vasild commented:
I am ok with 96571b3 (ACK-ed above). Feel free to ignore the idea or postpone it.

laanwj commented:
Re-ACK 96571b3d4cb4cda0fd3d5a457ae4a12f615de82b

jonatack commented:
@hebasto I am writing test coverage, are you open to bringing it into this PR or do you consider it a follow-up?

jonatack commented:
Oops, sorry.

laanwj commented:
@jonatack Though it would have been nice to include more testing here, let's leave that for a follow-up PR. this is a good self-contained change, and now we can move on to the other PRs based on it.

jonatack commented:
Ok. I asked because it's difficult to get follow-up test coverage reviewed and merged on its own, I already have a few PRs open like that which I'll probably close.

SimonVrouwe commented:
@hebasto Hello, I ended up here in to find reasons why my node (`bitcoind` v0.21.1) is listening on port 8334 even when `listenonion=0` is set. Reading the discussion nin this PR:

> @vasild
> 
...
> > ### With this PR
> > 
> > * If no `-bind` is given then we will bind on `0.0.0.0:8333`, `[::]:8333` and `127.0.0.1:8334`.
> > * If `-bind=1.2.3.4:8333` is given then we will bind on `1.2.3.4:8333` and on `127.0.0.1:8334` (is this ok?).
> > * If `-bind=1.2.3.4:8333 -bind=1.2.3.4:8334=onion` is given then we will bind on `1.2.3.4:8333` and `1.2.3.4:8334`.
> 
> The node will not bind on `127.0.0.1:8334` if non-default `listenonion=0` is provided.

@hebasto That last statement appears incorrect (at least in v0.21.1), with `listenonion=0` the node still binds to `127.0.0.1:8334`. I ended up here in search for reasons why the `onion_binds.push_back` is not placed inside the `if (args.GetBoolArg("-listenonion", DEFAULT_LISTEN_ONION))`

https://github.com/bitcoin/bitcoin/blob/2b5563bb1e6819bf46f561d94213d5fb32751359/src/init.cpp#L1725-L1740
